### DISCUSSION TITLE
Memory and CPU Efficient Triple Updates

### DISCUSSION THREAD

**Discussion by atodkar** [dotnetrdf/dotnetrdf#362] [Q&A]
I am using dotnetRDF in a constraint device and want to frequently update Triples. I see there is ExecuteUpdate which helps execute Update SPARQL on Triples. There is also another approach where I simply keep an index of Triples and Retract and Assert that particular triple for an update.

`
graph.Retract(triple);
var updated = new Triple(s, p, graph.CreateLiteralNode(val));
graph.Assert(updated);
`

Which one will be more efficient? How about Transactional behavior? Any help on benchmarking or past experiences will be helpful.

**Comment by kal** [comment]
Hi Anand,

If your use-case is this simple, then doing the Retract and Assert will be more efficient. The SPARQL Update implementation will be using the same methods under the hood, and so by doing them directly in your code you avoid the overhead of the SPARQL parser. 

Hope that helps!

**Reply by atodkar** [reply]
Thanks for your response. I am facing challenge to make Retract and Assert a transactional behavior. Problem is, I have several clients also simultaneously querying using ExecuteQuery. There is inconsistent behavior while we do simultanueous Updates and Query. Not sure if I need do this in a mutually exclusive way?

**Reply by atodkar** [reply]
By the way, do we have implementation of ITransactionalGraph interface? looks like this will be helpful. Also will this interface be able to create InMemory Graph?

**Reply by kal** [reply]
There is a wrapper graph implementation [`VDS.RDF.ThreadSafeGraph`](https://www.dotnetrdf.org/api/html/T_VDS_RDF_ThreadSafeGraph.htm) which essentially wraps all of the calls to the 'IGraph' methods in a 'ReaderWriterLockSlim' - so this could help, but it doesn't give your update operation isolation as the delete and assert would be two separate write operations. You provide the wrapper with the underlying graph that gets updated, so you can use this to add thread-safety to a simple in-memory 'Graph'.

There is another wrapper, ['VDS.RDF.GraphPersistenceWrapper'](https://www.dotnetrdf.org/api/html/T_VDS_RDF_GraphPersistenceWrapper.htm) which implements the `ITransactionalGraph` interface over an underlying graph that you provide (so again you can use this to extend the basic in-memory `Graph`).

However, there isn't a default implementation that does both things! But I think you could use the code from these implementations to put together your own `ThreadSafeGraphPersistenceWrapper` implementation that uses the lock pattern of `ThreadSafeGraph` over the `ITransactionalGraph` implementation in `GraphPersistenceWrapper`. That might be useful for other folks too, so if you feel like taking this on and contributing it back to the project we would welcome that!

**Reply by atodkar** [reply]
Thank you @kal appreciated. Let me see and try this. By the way, while searching I also got to know about InMemoryManager.UpdateGraph. Looks like this also is doing Retract and Assert internally. Is this of any use? Or if it makes more sense to make InMemoryManager.UpdateGraph transactional?

**Reply by kal** [reply]
I'd forgotten about `InMemoryManager`! I'd say that it implementing the IStorageServer interface is probably only really worth doing if you want to think about porting code from running in-memory to running against a remote triple store - it is really there to provide a route from testing code in-memory to then working with a store such as Virtuoso or Jena, and it means implementing a lot of graph management operations which might not be necessary in your use-case?

**Reply by atodkar** [reply]
Yes you are right about InMemoryManager, looks like extending ThreadSafeGraph and implementing Transactional behavior makes more sense.

I was looking at GraphPersistenceWrapper implementation and see its indeed using Assert and Retract internally. Flush is NoOp and Discard is indeed doing reverse operation. This is not really transactional because if I am doing ExecuteQuery (SELECT operation) in parallel thread, I get inconsistent results. Do you know how we can make it Transactional in the sense that on Flush of results the GraphPersistenceWrapper should actually update Graph? I see one way is to use lock on Graph object while doing Update operation. Would that be of big performance penalty? I am in the middle of implementing new class and naming it as `ThreadSafeGraphPersistenceWrapper` for in memory store. 

Let me know if you recommend something else.

**Reply by kal** [reply]
Ah yes, you are right. I think to get isolated reads what you would need to change it to record the actions without performing them and then have Flush() do the work of updating the underlying graph. So in Assert/Retract instead of asserting/retracting the triple in the underlying graph, just add the TriplePersistenceAction to the _actions list, then in Flush you can run through the actions applying them to the underlying graph or in Discard you just clear _actions. Does that make sense?

**Reply by atodkar** [reply]
Yes and then graph will be locked while doing flush operation. I think this makes sense.

**Reply by kal** [reply]
I just realised that there is still potentially another issue. 

If you have multiple writers then there is a problem because all transactions share the same list of updates. So potenially writerA could start an update, writerB starts an update before A completes and then when writerA flushes, both updates from writerA and from writerB could get applied. I'm not sure that there is a way around this with the API as it currently is because there is no call to start a transaction. 

That means that even with these changes, in your code you will need something to ensure that there is only ever a single writer (maybe a lock that a writer needs to get before starting their update?). If you do that, then you might want to look at whether it would be easier to just have a read/write lock in your data access layer and continue using the basic in-memory graph.

**Reply by atodkar** [reply]
Yes right.
I think the User on any operation (ExecuteQuery or Assert/Retract) should create separate object of  ThreadSafeGraphPersistenceWrapper. This will keep list of `TriplePersistenceAction` separate for each operation. Will use static ReadWriteLockSlim which will make sure all operations are threadsafe regardless of objects. 

 Do you think this approach will work?
